v3/newrelic/code_level_metrics.go

```diff
@@ -171,13 +343,58 @@ func ThisCodeLocation(skipLevels ...int) *CodeLocation {
 	return &loc
 }
 
+//
+// ThisCodeLocation works identically to the stand-alone ThisCodeLocation function,
+// in that it determines the souce code location from whence it was called, returning
+// a pointer to a CodeLocation value which represents that location. However,
+// unlike the stand-alone function, this stores the result in the CachedCodeLocation receiver;
+// thus, subsequent invocations of ThisCodeLocation for the same receiver will result in
+// immediately repeating the value obtained from the first
+// invocation.
+//
+// This is thread-safe and is intended to allow the same code to run in multiple
+// concurrent goroutines without needlessly recalculating the location of the
+// caller.
+//
+func (c *CachedCodeLocation) ThisCodeLocation(skiplevels ...int) *CodeLocation {
+	var skip int
+
+	if len(skiplevels) > 0 {
+		skip = skiplevels[0]
+	}
+
+	c.once.Do(func() {
+		// add 4 skip levels to compensate for the internal calls used to get here.
```

Comment:
This is contingent upon the implementation of the sync package, which could be a problem.

Reply:
What issues do you see with the sync package? I thought that was a fairly well-defined part of the standard library at this point. But if there are issues we could explicitly document expectations around that or do something to mitigate the issues.

Reply:
The fact that there are 4 levels is an implementation detail. If the Go authors refactor the way `sync.Once.Do` works internally it could change.

Reply:
Oh, that. Yes, that's on my mind as well. I'm afraid there may need to be some more logic in there to do something more intelligent than blindly skipping a number of levels, like skipping over things that belong to the sync package.

Reply:
The code just pushed includes a new implementation that more intentionally steps around the internal agent code and calls out to the sync library and back rather than just skipping a number of calls, which should mitigate this issue.